src/shims/lock.c

```diff
@@ -600,33 +648,42 @@ _dispatch_unfair_lock_unlock_slow(dispatch_unfair_lock_t dul, dispatch_lock cur)
 #pragma mark - gate lock
 
 void
-_dispatch_gate_wait_slow(dispatch_gate_t dgl, dispatch_lock value,
-		dispatch_lock_options_t flags)
+_dispatch_once_wait(dispatch_once_gate_t dgo)
 {
 	dispatch_lock self = _dispatch_lock_value_for_self();
-	dispatch_lock old_value, new_value;
+	uintptr_t old_v, new_v;
+	dispatch_lock *lock = &dgo->dgo_gate.dgl_lock;
 	uint32_t timeout = 1;
 
 	for (;;) {
-		os_atomic_rmw_loop(&dgl->dgl_lock, old_value, new_value, acquire, {
-			if (likely(old_value == value)) {
-				os_atomic_rmw_loop_give_up_with_fence(acquire, return);
+		os_atomic_rmw_loop(&dgo->dgo_once, old_v, new_v, relaxed, {
+			if (likely(old_v == DLOCK_ONCE_DONE)) {
+				os_atomic_rmw_loop_give_up(return);
+			}
+#if DISPATCH_ONCE_USE_QUIESCENT_COUNTER
+			if (DISPATCH_ONCE_IS_GEN(old_v)) {
+				os_atomic_rmw_loop_give_up({
+					os_atomic_thread_fence(acquire);
+					return _dispatch_once_mark_done_if_quiesced(dgo, old_v);
+				});
 			}
-			new_value = old_value | DLOCK_WAITERS_BIT;
-			if (new_value == old_value) os_atomic_rmw_loop_give_up(break);
+#endif
+			new_v = old_v | (uintptr_t)DLOCK_WAITERS_BIT;
+			if (new_v == old_v) os_atomic_rmw_loop_give_up(break);
 		});
-		if (unlikely(_dispatch_lock_is_locked_by(old_value, self))) {
+		if (unlikely(_dispatch_lock_is_locked_by((dispatch_lock)old_v, self))) {
 			DISPATCH_CLIENT_CRASH(0, "trying to lock recursively");
 		}
 #if HAVE_UL_UNFAIR_LOCK
-		_dispatch_unfair_lock_wait(&dgl->dgl_lock, new_value, 0, flags);
+		_dispatch_unfair_lock_wait(lock, (dispatch_lock)new_v, 0,
+				DLOCK_LOCK_NONE);
 #elif HAVE_FUTEX
-		_dispatch_futex_wait(&dgl->dgl_lock, new_value, NULL, FUTEX_PRIVATE_FLAG);
+		_dispatch_futex_wait(lock, (dispatch_lock)new_v, NULL,
+				FUTEX_PRIVATE_FLAG);
 #else
-		_dispatch_thread_switch(new_value, flags, timeout++);
+		_dispatch_thread_switch(new_v, flags, timeout++);
```

Comment:
`flags` was removed here

Reply:
Good point. I think 0 would work for flags (and in fact I don't think the value will be used, at least not on Windows).

Reply:
Yeah, it won't be used on Windows, but android I think might.